File: apps/web/src/types/api/excerpt-assets.ts
export type ExcerptAssetStateDto =
  | "all"
  | "favorite"
  | "highlight"
  | "note"
  | "insight";

export type ExcerptAnchorTypeDto = "sentence" | "text_range" | "multi_text";

export interface ExcerptInsightDto {
  id: string;
  type: "grammar" | "sentence" | "term" | "logic" | "interpretation" | "summary";
  label: string;
  title: string;
  detail?: string | null;
}

export interface ExcerptAssetSegmentDto {
  paragraph_id?: string | null;
  sentence_id: string;
  selected_text: string;
  start_offset: number;
  end_offset: number;
  text_hash: string;
}

export interface ExcerptAssetItemDto {
  target_key: string;
  anchor_type: ExcerptAnchorTypeDto;
  sentence_id?: string | null;
  selected_text: string;
  translation?: string | null;
  start_offset?: number | null;
  end_offset?: number | null;
  text_hash?: string | null;
  segments: ExcerptAssetSegmentDto[];
  updated_at: string;
  is_favorited: boolean;
  is_highlighted: boolean;
  is_noted: boolean;
  annotation_id?: string | null;
  annotation_type?: string | null;
  annotation_color?: string | null;
  note?: string | null;
  insights: ExcerptInsightDto[];
}

export interface ExcerptAssetGroupDto {
  record_id: string;
  client_record_id?: string | null;
  title: string;
  subtitle?: string | null;
  updated_at: string;
  asset_count: number;
  items: ExcerptAssetItemDto[];
}

export interface ExcerptAssetsResponseDto {
  groups: ExcerptAssetGroupDto[];
  total_assets: number;
  total_groups: number;
  page: number;
  limit: number;
}

File: services/api/app/schemas/excerpt_assets.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


ExcerptAssetState = Literal["all", "favorite", "highlight", "note", "insight"]
ExcerptAnchorType = Literal["sentence", "text_range", "multi_text"]
ExcerptInsightType = Literal[
    "grammar",
    "sentence",
    "term",
    "logic",
    "interpretation",
    "summary",
]


class ExcerptInsight(BaseModel):
    id: str
    type: ExcerptInsightType
    label: str
    title: str
    detail: str | None = None


class ExcerptAssetSegment(BaseModel):
    paragraph_id: str | None = None
    sentence_id: str
    selected_text: str
    start_offset: int
    end_offset: int
    text_hash: str


class ExcerptAssetItem(BaseModel):
    target_key: str
    anchor_type: ExcerptAnchorType
    sentence_id: str | None = None
    selected_text: str
    translation: str | None = None
    start_offset: int | None = None
    end_offset: int | None = None
    text_hash: str | None = None
    segments: list[ExcerptAssetSegment] = Field(default_factory=list)
    updated_at: str
    is_favorited: bool
    is_highlighted: bool
    is_noted: bool
    annotation_id: str | None = None
    annotation_type: str | None = None
    annotation_color: str | None = None
    note: str | None = None
    insights: list[ExcerptInsight] = Field(default_factory=list)


class ExcerptAssetGroup(BaseModel):
    record_id: str
    client_record_id: str | None = None
    title: str
    subtitle: str | None = None
    updated_at: str
    asset_count: int
    items: list[ExcerptAssetItem] = Field(default_factory=list)


class ExcerptAssetsResponse(BaseModel):
    groups: list[ExcerptAssetGroup]
    total_assets: int
    total_groups: int
    page: int
    limit: int
